src/services/earn/earn-service.ts

```diff
@@ -0,0 +1,262 @@
+import { IPermit2Service, PermitData } from '@services/permit2';
+import { IQuoteService, QuoteRequest } from '@services/quotes';
+import { IProviderService } from '@services/providers';
+import { MULTICALL_CONTRACT } from '@services/providers/utils';
+import { IAllowanceService } from '@services/allowances';
+
+export class EarnService implements IEarnService {
+  constructor(
+    private readonly permit2Service: IPermit2Service,
+    private readonly quoteService: IQuoteService,
+    private readonly providerService: IProviderService,
+    private readonly allowanceservice: IAllowanceService
+  ) {}
+
+  async buildCreatePositionTx({
+    chainId,
+    strategyId,
+    owner,
+    permissions,
+    strategyValidationData,
+    misc,
+    deposit,
+  }: CreateEarnPositionParams): Promise<BuiltTransaction> {
+    let depositInfo: { token: Address; amount: bigint; value: bigint };
+    if ('token' in deposit) {
+      const amount = BigInt(deposit.amount);
+      depositInfo = { token: deposit.token, amount, value: isSameAddress(deposit.token, Addresses.NATIVE_TOKEN) ? amount : 0n };
+    } else {
+      depositInfo = { token: deposit.permitData.token, amount: BigInt(deposit.permitData.amount), value: 0n };
+    }
+
+    const { needsSwap, asset } = await this.checkIfNeedsSwap({ chainId, strategyId, depositToken: depositInfo.token });
+
+    if ('token' in deposit && !needsSwap) {
+      // If don't need to use Permit2, then just call the vault
+      return {
+        to: EARN_VAULT.address(chainId),
+        data: encodeFunctionData({
+          abi: vaultAbi,
+          functionName: 'createPosition',
+          args: [
+            strategyId,
+            depositInfo.token as ViemAddress,
+            depositInfo.amount,
+            owner as ViemAddress,
+            permissions.map(({ operator, permissions }) => ({
+              operator: operator as ViemAddress,
+              permissions: permissions.map(mapPermission),
+            })),
+            strategyValidationData ?? '0x',
+            misc ?? '0x',
+          ],
+        }),
+        value: isSameAddress(depositInfo.token, Addresses.NATIVE_TOKEN) ? depositInfo.amount : undefined,
+      };
+    }
+    // If we get to this point, then we'll use the Companion for the deposit
+    const calls: Hex[] = [];
+
+    // Handle take from caller (if necessary)
+    const recipient = needsSwap ? COMPANION_SWAPPER_CONTRACT.address(chainId) : EARN_VAULT_COMPANION.address(chainId);
+    if ('permitData' in deposit) {
+      calls.push(buildTakeFromCallerWithPermit(deposit.permitData, deposit.signature, recipient));
+    } else if (!isSameAddress(depositInfo.token, Addresses.NATIVE_TOKEN)) {
+      calls.push(buildTakeFromCaller(depositInfo.token, depositInfo.amount, recipient));
+    }
+
+    const depositToken = (needsSwap ? asset : depositInfo.token) as ViemAddress;
+
+    const [, allowances] = await Promise.all([
+      async () => {
```

Comment:
This works? Wow. TIL

Reply:
Not really, it needs one `()` at the end